### PR TITLE
Tag searcher CLI executable

### PR DESCRIPTION
**Pull request by okofish**
## Changes

* Add an executable target (enabled with the `POST_TAG_SYSTEM_BUILD_CLI` CMake option) that builds a command-line binary for running searches over ranges of initial conditions
* Add a Dockerfile for building the CLI
* Add `PostTagFileReader` and `PostTagFileWriter` utility base classes for parsing/serializing the ad-hoc binary data file formats that will be used to store "crib" sequences (known checkpoints), initial conditions, and evaluation results.
* Add a `PostTagCribFile` file format struct and accompanying reader class
* Add a `PostTagResultFile` file format struct and accompanying writer class

## Comments

* The executable is mostly scaffolding right now, to be filled in once the PostTagSearcher API (#14) is implemented.
* [`Boost.Program_options`](https://www.boost.org/doc/libs/1_67_0/doc/html/program_options.html) is required for command-line option parsing. Boost static/shared libraries and headers can be found in many package managers.
* The executable can be built with static linking by enabling the `POST_TAG_SYSTEM_CLI_STATIC_BUILD` CMake option.

---
This change is [Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)

### CODE REVIEW

**Review by maxitg — COMMENTED** [review]
> Reviewed 18 of 21 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: all files reviewed, 6 unresolved discussions (waiting on @okofish)

___
*[CLI/arguments.cpp, line 54 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUduRr_CKy67ZRRlpxm:-MUduRr_CKy67ZRRlpxn:b3g40wq) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L54)):*
> ```cpp
>     ("cribfile,f",    po::value<std::string>()->value_name("path"),
>       "Path to crib file (list of known sequences)")
>     ("cribsize,b",    po::value<uint32_t>()->value_name("size")
> ```

How is this going to be used? Doesn't the `cribfile` contain that already?
___
*[CLI/arguments.cpp, line 62 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUduiO_86Kyx6a7AeWq:-MUduiO_86Kyx6a7AeWr:btdf0qq) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L62)):*
> ```cpp
>     ("initstart,s",   po::value<uint64_t>()->value_name("start"),
>       "Starting initial condition")
>     ("initcount,n",   po::value<uint64_t>()->default_value(1)->value_name("count")
> ```

Good idea. This will prevent people from being confused about `end` vs. `end - 1`.
___
*[CLI/arguments.cpp, line 66 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdwqT85ifGbk7nQb0k:-MUdwqT85ifGbk7nQb0l:b92vwxz) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L66)):*
> ```cpp
>       "Number of initial conditions")
>     ("initoffset,e",  po::value<uint64_t>()->default_value(0)->value_name("offset"),
>       "Initial condition offset (for use with zero-indexed array jobs)")
> ```

This description is not very clear to me. I think the comment in `main.cpp` is better. Or maybe it can say something like "shift all inits by offset * count".
___
*[CLI/main.cpp, line 23 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdwTMO6B-JpJUgMgCR:-MUdwTMO6B-JpJUgMgCS:bolfnwa) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/main.cpp#L23)):*
> ```cpp
> }
> 
> void print_bits(std::vector<bool> bits) {
> ```

`const std::vector<bool>& bits`
___
*[CLI/main.cpp, line 56 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdz-we-MKvp3D8YpJT:-MUdz-we-MKvp3D8YpJU:bptxq0j) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/main.cpp#L56)):*
> ```cpp
>       auto checkpoint = checkpoint_spec.states[i];
>       print_bits(checkpoint.tape);
>       printf(" - %u\n", checkpoint.headState);
> ```

We probably need a `--verbose` flag which would disable this, and also the printing of the results.
___
*[CLI/files/PostTagCribFile.cpp, line 7 at r2](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUe0smU9MxhusZQcsEq:-MUe0smU9MxhusZQcsEr:b-uk0kuz) ([raw file](https://github.com/maxitg/posttagsystem/blob/cef0bcd73424cdafe26d0d98d203bf508f708aec/CLI/files/PostTagCribFile.cpp#L7)):*
> ```cpp
> using PostTagSystem::PostTagState;
> 
> PostTagCribFile PostTagCribFileReader::read_file() {
> ```

It would be nice to add a googletest for this using `TestDatafiles/crib1.postcrib`. (And the same for reading/writing other files.)
___
*[TestDatafiles/crib1.crib, line 0 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdpDKjASjl0FGy5dRc:-MUdpDKjASjl0FGy5dRd:bigvqqc) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/TestDatafiles/crib1.crib#L0)):*
Nit: It might be useful to add cases with tapes over 8 bits for testing.

**Review by okofish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: all files reviewed, 6 unresolved discussions (waiting on @maxitg and @okofish)

___
*[CLI/arguments.cpp, line 54 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUduRr_CKy67ZRRlpxm:-MUeASbpEayroZd94y64:b-vha5n) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L54)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

How is this going to be used? Doesn't the `cribfile` contain that already?
</blockquote></details>

Good catch - this was left over from the initial spec; I'll remove it. Each sequence in `cribfile` is indeed prefixed with its length.
___
*[CLI/main.cpp, line 56 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdz-we-MKvp3D8YpJT:-MUeBBx1-a9PysTzDjKc:b3px3bs) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/main.cpp#L56)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

We probably need a `--verbose` flag which would disable this, and also the printing of the results.
</blockquote></details>

This (and most of `main.cpp`) is just scaffolding for testing; once the searcher API is integrated there won't be anything printed except maybe a brief summary at the very end of the job.

It might be useful to have a `--verbose` flag that causes a status line to be printed every N seconds (or every N inits), but I think that would have to happen via some status callback that I would pass into your `PostTagSearcher`.
___
*[CLI/files/PostTagCribFile.cpp, line 7 at r2](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUe0smU9MxhusZQcsEq:-MUeBxPt4Yeq6vVoVFcn:b9eth56) ([raw file](https://github.com/maxitg/posttagsystem/blob/cef0bcd73424cdafe26d0d98d203bf508f708aec/CLI/files/PostTagCribFile.cpp#L7)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

It would be nice to add a googletest for this using `TestDatafiles/crib1.postcrib`. (And the same for reading/writing other files.)
</blockquote></details>

That's a good idea. I'll do that once the file formats are more solidified.
___
*[TestDatafiles/crib1.crib, line  at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdpDKjASjl0FGy5dRc:-MUeC9hbDF7FuBcd6fpt:bjoej4s) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/TestDatafiles/crib1.crib#L)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Nit: It might be useful to add cases with tapes over 8 bits for testing.
</blockquote></details>

Definitely; will do.

**Review by okofish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: 18 of 24 files reviewed, 5 unresolved discussions (waiting on @maxitg)

___
*[CLI/arguments.cpp, line 54 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUduRr_CKy67ZRRlpxm:-MUeEYKf3Y9aSWC1Ghdj:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L54)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

Good catch - this was left over from the initial spec; I'll remove it. Each sequence in `cribfile` is indeed prefixed with its length.
</blockquote></details>

Done.
___
*[CLI/arguments.cpp, line 66 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdwqT85ifGbk7nQb0k:-MUeDc7w5uz9SW2SDBbM:b-i6079j) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L66)):*
Changed to:
> Shifts starting condition by offset * count (for use with zero-indexed array jobs)
___
*[CLI/main.cpp, line 23 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdwTMO6B-JpJUgMgCR:-MUeCnWDBkgzbk5sZvmK:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/main.cpp#L23)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

`const std::vector<bool>& bits`
</blockquote></details>

Done.

**Review by okofish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: 18 of 24 files reviewed, 5 unresolved discussions (waiting on @maxitg)

___
*[CLI/arguments.cpp, line 66 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdwqT85ifGbk7nQb0k:-MUeEaR4B6BRyhUpfMqw:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/arguments.cpp#L66)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

Changed to:
> Shifts starting condition by offset * count (for use with zero-indexed array jobs)
</blockquote></details>

Done.

**Review by maxitg — COMMENTED** [review]
> Reviewed 4 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: 22 of 24 files reviewed, 1 unresolved discussion (waiting on @maxitg)

___
*[CLI/main.cpp, line 56 at r1](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUdz-we-MKvp3D8YpJT:-MUeEfAfACy7KMb299K-:b-ovwaw7) ([raw file](https://github.com/maxitg/posttagsystem/blob/dfc9e65120a83185256dfb0198c8efe32353e699/CLI/main.cpp#L56)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

This (and most of `main.cpp`) is just scaffolding for testing; once the searcher API is integrated there won't be anything printed except maybe a brief summary at the very end of the job.

It might be useful to have a `--verbose` flag that causes a status line to be printed every N seconds (or every N inits), but I think that would have to happen via some status callback that I would pass into your `PostTagSearcher`.
</blockquote></details>

Good idea.

File: libPostTagSystem/PostTagSearcher.hpp [hunk 1]
```diff
@@ -12,15 +12,18 @@ class PostTagSearcher {
...
+    Terminated = 2,
+    ReachedCycle = 3,
+    ReachedKnownCheckpoint = 4,
+    MaxTapeLengthExceeded = 5,
+    MaxEventCountExceeded = 6,
+    TimeConstraintExceeded = 7,
+    NotEvaluated = 8,
+
+    // result file spec requires the reason to fit in 5 bits
+    MAX_DO_NOT_EXCEED = 31
```
Comment (okofish) [inline libPostTagSystem/PostTagSearcher.hpp:26]: Calling this out as it's my one modification to the library code.

**Review by maxitg — CHANGES_REQUESTED** [review]
> Reviewed 24 of 32 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r10, 5 of 5 files at r11.
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: all files reviewed, 11 unresolved discussions (waiting on @maxitg and @okofish)

___
*[.circleci/config.yml, line 189 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjXZxl1XzMUEKJavYj:-MUjXZxl1XzMUEKJavYk:b-9frfpg) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/.circleci/config.yml#L189)):*
> ```YAML
>           path: ./LibraryResources/
> 
>   windows-cli-build:
> ```

Wouldn't it make more sense to build CLI in the above job, given that the setup is much more expensive than the build itself? Plus, we'll avoid code duplication for the "Install CMake" step.
___
*[CLI/arguments.cpp, line 68 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjgZyI73jsqVcnAwUq:-MUjgZyI73jsqVcnAwUr:b4khvda) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/arguments.cpp#L68)):*
> ```cpp
>     ("cribfile,f",    po::value<std::string>()->value_name("path.postcrib"),
>       "Path to crib file (list of known sequences)")
>     ("initsize,l",    po::value<uint16_t>()->default_value(30)->value_name("size")
> ```

I think `initsize` should be a required parameter. It does not really make much sense to specify the range without it.
___
*[CLI/main.cpp, line 120 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjbbt9ANk1jEtTSoZT:-MUjbbt9ANk1jEtTSoZU:b-6gnv5j) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/main.cpp#L120)):*
> ```cpp
>     start += count * offset;
> 
>     std::cout << boost::format("Evaluating %u initial conditions, starting at %u...\n") % count % start;
> ```

"%u initial tapes", there are 3 * %u initial conditions.
___
*[CLI/main.cpp, line 124 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjbK643tONwZPafuov:-MUjbK643tONwZPafuow:b-mam52b) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/main.cpp#L124)):*
> ```cpp
> 
>     try {
>       results = searcher.evaluateRange(tape_length, start, start + count, eval_params);
> ```

It's a bit weird that we are evaluating `3 * initcount` inits. Would it make more sense to rename the flags to `--inittapestart` and `--inittapecount`?
___
*[FileFormats/CribFile.wl, line 90 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjkg6f9_1CDQhCfzuW:-MUjkg6f9_1CDQhCfzuX:b-58hvc9) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/CribFile.wl#L90)):*
> ```Mathematica
>     version,
>     stream,
>     tape -> 0
> ```

I don't think the default head state of zero makes much sense. Head state is equally as important as the tape.
___
*[FileFormats/InitFile.wl, line 99 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjlB3XB2oRMK_N5OBr:-MUjlB3XB2oRMK_N5OBs:b-qdzzn) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/InitFile.wl#L99)):*
> ```Mathematica
>     version,
>     stream,
>     tape -> 0
> ```

Same here wrt the default head state.
___
*[FileFormats/PostTagFileFormats.wl, line 1 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjlSZ89yBEOi22KHCJ:-MUjlSZ89yBEOi22KHCK:b-vkta6e) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/PostTagFileFormats.wl#L1)):*
> ```Mathematica
> BeginPackage["PostTagFileFormats`"]
> ```

Why not put all this into the main package (in the Kernel directory)? People like Stephen will probably want to use this, and they won't be able to unless we put it in the paclet (or make another one). Plus, this requires `SetDirectory` as without it the following does not work:

```wl
Get["~/git/PostTagSystem/FileFormats/PostTagFileFormats.wl"]
```
___
*[FileFormats/ResultFile.wl, line 133 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjkKhW227mqXxGNmx1:-MUjkKhW227mqXxGNmx2:b7yvoif) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/ResultFile.wl#L133)):*
> ```Mathematica
> (* Register import converter *)
> 
> ImportExport`RegisterImport[
> ```

Is it possible to make it automatically recognize `.postresult` extension?
___
*[libPostTagSystem/PostTagSearcher.hpp, line 26 at r7](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUg-251EEGPh9ghcYIH-r7-26:-MUjoogjDVpGJ5x1swwt:b-etsth8) ([raw file](https://github.com/maxitg/posttagsystem/blob/15c3f07aa403a95121b1b295d3249d14b86f1b90/libPostTagSystem/PostTagSearcher.hpp#L26)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

Calling this out as it's my one modification to the library code.
</blockquote></details>

Sounds good. 31 should be sufficient.
___
*[FileFormats/crib2.postcrib, line 0 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjUsQb7qjGQxIEtUIl:-MUjUsQb7qjGQxIEtUIm:bwaxihi) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/crib2.postcrib#L0)):*
Should this file be in the `SampleFiles` directory?

**Review by okofish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: 29 of 33 files reviewed, 11 unresolved discussions (waiting on @maxitg and @okofish)

___
*[.circleci/config.yml, line 189 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjXZxl1XzMUEKJavYj:-MUjujmS6PlkNWt8i0E7:bp5plz0) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/.circleci/config.yml#L189)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Wouldn't it make more sense to build CLI in the above job, given that the setup is much more expensive than the build itself? Plus, we'll avoid code duplication for the "Install CMake" step.
</blockquote></details>

Makes sense; I'll try to get this working via `buildLibraryResources.sh` (but only on Windows for now, since we don't install Boost on macOS). This may take a few iterations of commit/check CircleCI/commit again.
___
*[CLI/arguments.cpp, line 68 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjgZyI73jsqVcnAwUq:-MUjuicf9yrtGVlocLBy:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/arguments.cpp#L68)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

I think `initsize` should be a required parameter. It does not really make much sense to specify the range without it.
</blockquote></details>

Done.
___
*[CLI/main.cpp, line 120 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjbbt9ANk1jEtTSoZT:-MUjuxFYA9zKTZgphkCq:b-b8u25i) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/main.cpp#L120)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

"%u initial tapes", there are 3 * %u initial conditions.
</blockquote></details>

Changed to `%u initial condition tapes`.
___
*[CLI/main.cpp, line 124 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjbK643tONwZPafuov:-MUjv77IEy9q8I9sfBCU:b-z1sjq4) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/CLI/main.cpp#L124)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

It's a bit weird that we are evaluating `3 * initcount` inits. Would it make more sense to rename the flags to `--inittapestart` and `--inittapecount`?
</blockquote></details>

Rather than make the flags longer I changed the descriptions to include "tape"; let me know if this isn't enough.
___
*[FileFormats/CribFile.wl, line 90 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjkg6f9_1CDQhCfzuW:-MUjvhua2npsNGF51Fd0:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/CribFile.wl#L90)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

I don't think the default head state of zero makes much sense. Head state is equally as important as the tape.
</blockquote></details>

Done.
___
*[FileFormats/InitFile.wl, line 99 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjlB3XB2oRMK_N5OBr:-MUjvhKkADw2zMgByrp3:b-896fix) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/InitFile.wl#L99)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Same here wrt the default head state.
</blockquote></details>

Done.
___
*[FileFormats/PostTagFileFormats.wl, line 1 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjlSZ89yBEOi22KHCJ:-MUjvkM5D6dI865TwTmn:bh6gvmw) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/PostTagFileFormats.wl#L1)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Why not put all this into the main package (in the Kernel directory)? People like Stephen will probably want to use this, and they won't be able to unless we put it in the paclet (or make another one). Plus, this requires `SetDirectory` as without it the following does not work:

```wl
Get["~/git/PostTagSystem/FileFormats/PostTagFileFormats.wl"]
```
</blockquote></details>

I thought about this previously but I don't know how to use the "new" package format - can I safely just drop these files in the Kernel directory without modifications or do I have to rewrite them in the new style?

I would be OK with making another paclet in this folder; at least I know how to do that.
___
*[FileFormats/ResultFile.wl, line 133 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjkKhW227mqXxGNmx1:-MUjwNSP8gdCti4LhId1:bxmc38z) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/ResultFile.wl#L133)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Is it possible to make it automatically recognize `.postresult` extension?
</blockquote></details>

AFAIK only via this: https://resources.wolframcloud.com/FunctionRepository/resources/RegisterFormat

I could put in a utility function like `RegisterPostTagFileFormats[]` that calls that ResourceFunction. I'm not sure how I would feel about copying the big chunk of mysterious internal code from that ResourceFunction directly into these files. Then again, what's the worst that could happen...
___
*[FileFormats/crib2.postcrib, line  at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjUsQb7qjGQxIEtUIl:-MUjvPea3VUyZvGvslt1:b-fwcmoa) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/crib2.postcrib#L)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Should this file be in the `SampleFiles` directory?
</blockquote></details>

I don't even remember what that was for; deleted.

**Review by maxitg — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r12.
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16)* status: all files reviewed, 4 unresolved discussions (waiting on @maxitg and @okofish)

___
*[.circleci/config.yml, line 189 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjXZxl1XzMUEKJavYj:-MUk3fElAh4yDDdCHNQc:b-dwittz) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/.circleci/config.yml#L189)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

Makes sense; I'll try to get this working via `buildLibraryResources.sh` (but only on Windows for now, since we don't install Boost on macOS). This may take a few iterations of commit/check CircleCI/commit again.
</blockquote></details>

Why don't you just run your build step after `buildLibraryResources`? I think that would be more reliable, and it wouldn't be any slower because of caching.
___
*[FileFormats/PostTagFileFormats.wl, line 1 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjlSZ89yBEOi22KHCJ:-MUk5_zW5_eNk5O2hS9J:bozhu4f) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/PostTagFileFormats.wl#L1)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

I thought about this previously but I don't know how to use the "new" package format - can I safely just drop these files in the Kernel directory without modifications or do I have to rewrite them in the new style?

I would be OK with making another paclet in this folder; at least I know how to do that.
</blockquote></details>

It's [quite simple](https://github.com/maxitg/SetReplace/blob/master/.github/CONTRIBUTING.md#wolfram-language-code) actually. You cannot just copy these files there, but all you need to change is add ``Package["PostTagSystem`"]`` at the top of every file, remove all `Begin/EndPackage` and `Begin/End`. And then, put `PackageExport["symbolName"]` for all public symbols and `PackageScope["symbolName"]` for all symbols used in other files. The package system loads all files automatically in alphabetical order (so you might need to put `A1$` in front of `Utility`).
___
*[FileFormats/ResultFile.wl, line 133 at r8](https://reviewable.io/reviews/maxitg/posttagsystem/16#-MUjkKhW227mqXxGNmx1:-MUk4P_O19KWA285HqW9:b28fnev) ([raw file](https://github.com/maxitg/posttagsystem/blob/33ce05580efb4be58bcea4bfce29adee487aa386/FileFormats/ResultFile.wl#L133)):*
<details><summary><i>Previously, okofish (Jesse Friedman) wrote…</i></summary><blockquote>

AFAIK only via this: https://resources.wolframcloud.com/FunctionRepository/resources/RegisterFormat

I could put in a utility function like `RegisterPostTagFileFormats[]` that calls that ResourceFunction. I'm not sure how I would feel about copying the big chunk of mysterious internal code from that ResourceFunction directly into these files. Then again, what's the worst that could happen...
</blockquote></details>

I don't think we should be calling resource functions here. It's too unreliable. But with lines like this, it really is quite mysterious:
```wl
{fmt, True, False, False, False, False, $ext, $mime, None, {}}
```
We can copy that resource function to a new file, but maybe we shouldn't worry about it for now, given that we don't have much time left.

**Review by maxitg — APPROVED** [review]
I'm going to approve now so that I don't block you. I think it would be nice to merge `FileFormats` to the main package, but it can be done in a separate PR.

> Reviewed 1 of 2 files at r13.
> *[Reviewable](https://reviewable.io/reviews/maxitg/posttagsystem/16#-:-MUkXOyq3B7HWQ-rAixr:b51avj5)* status: all files reviewed, 3 unresolved discussions (waiting on @maxitg and @okofish)

**Pull request merged**